### PR TITLE
[CPDEV-94111] - fix ctr flags for new containerd format

### PR DESCRIPTION
**Pull request by Imadzuma**
### Description
This PR fixes the bug, related to PR https://github.com/Netcracker/KubeMarine/pull/523:
If user specifies only new format configuration (without credentials), `hosts-dir` flag is not applied for ctr commands.
Example of configuration:
```yaml
services:
  kubeadm:
    imageRepository: registry.example:9090
  cri:
    containerdRegistriesConfig:
      registry.example:9090:
        host."http://registry.example:9090":
          capabilities: [ "pull", "resolve" ]
          skip_verify: true
```
As result ctr fails with pulling problem like:
```
failed to do request: Head \"https://registry.example:9090/v2/pause/manifests/3.9\": http: server gave HTTP response to HTTPS client" host="registry.example:9090"
ctr: failed to resolve reference "registry.example:9090/pause:3.9": failed to do request: Head "https://registry.example:9090/v2/pause/manifests/3.9": http: server gave HTTP response to HTTPS client
```
The root cause in ctr_flags.conf that saves ctr options per registry in old format, but --hosts-dir should be applied for any registry, if the new format is used. 

Fixes # (issue)

### Solution
* Added `*` rule in ctr_flags.conf, that means, that options should be applied for any registry;
* Fixed `etcdctl`, that can parse '*' rules in ctr_flags.conf;
* If `config-path` is enabled, `*=--hosts-dir <config_path>` is added to  ctr_flags.conf;
* Added patch to reinstall etcdctl;
* Fixed reconfigure_registry patch using deviated_cluster, because it's more safe way to compare configuration differences when `raw_inventory`;

### Test Cases

**TestCase 1**

Test Configuration:

- Hardware: 
- OS: 
- Inventory: containerd configuration like from description above

Steps:

1. Kubemarine install
2. Kubemarine check_paas

Results:

| Before | After |
| ------ | ------ |
| `etcd.health_status` failed with error: ctr: image "registry.example:9090/etcd:3.5.6-0": not found | `etcd.health_status` task works correctly |

### Checklist
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Integration CI passed
- [ ] Unit tests. If Yes list of new/changed tests with brief description
- [ ] There is no merge conflicts

#### Unit tests
Indicate new or changed unit tests and what they do, if any.

### CODE REVIEW

File: kubemarine/cri/containerd.py [hunk 1]
```diff
@@ -256,11 +256,12 @@ def configure_ctr_flags(group: NodeGroup) -> None:
         elif registry_auth.get('username'):
             options.append(f'--user {registry_auth["username"]}' +
                            f':{registry_auth["password"]}' if registry_auth.get("password") else '')
-        # Add hosts-dir, if it's presented
-        if config_path:
-            options.append(f'--hosts-dir {config_path}')
```
Comment (ilia1243) [inline kubemarine/cri/containerd.py:261]: Won't it be simpler if just iterate over `set().union(registry.get('mirrors', {}).keys(), registry.get('configs', {}).keys(), cluster.inventory.get("services", {}).get("cri", {}).get('containerdRegistriesConfig', {}))`, and return these lines? In this case we could avoid extra patch

File: kubemarine/core/defaults.py [hunk 2]
```diff
@@ -171,7 +171,8 @@ def apply_registry(inventory: dict, cluster: KubernetesCluster) -> dict:
             containerd_endpoints = ["%s://%s" % (protocol, registry_mirror_address)]
 
         old_format_result, _ = contains_old_format_properties(inventory)
-        if old_format_result:
+        # todo remove p3_reconfigure_registries after next release
+        if old_format_result or (cluster and not cluster.context.get('p3_reconfigure_registries', True)):
```
Comment (ilia1243) [inline kubemarine/core/defaults.py:175]: bool(cluster) is always True. Also I suppose the same check should be applied to all cases where enrichment has changed. I remember that we also changed enrichment in containerd.enrich_inventory

File: kubemarine/patches/p3_reconfigure_registries.py [hunk 3]
```diff
@@ -25,18 +25,18 @@ def __init__(self) -> None:
...
-        result, _ = contains_old_format_properties(raw_inventory)
-        if result:
-            logger.info("Found old format configuration, updating containerd configuration is not needed")
+
+        prev_cluster = res.create_deviated_cluster({
+            'p3_reconfigure_registries': False
+        })
+
+        if 'containerdRegistriesConfig' not in cluster.inventory['services']['cri'] or \
+                'containerdRegistriesConfig' in prev_cluster.inventory['services']['cri']:
```
Comment (ilia1243) [inline kubemarine/patches/p3_reconfigure_registries.py:38]: Not sure that I understand these conditions. Is the second condition can really be True in some cases? Consider that 'containerdRegistriesConfig' cannot not be specified for old inventories. Maybe the first condition is enough and create_deviated_cluster is not necessary at all. Alternatively, the result of old and new enrichment can be compared by just `cluster.inventory['services']['cri'] == prev_cluster.inventory['services']['cri']`

File: kubemarine/cri/containerd.py (hunk 1 above)
Comment (ilia1243) [reply kubemarine/cri/containerd.py:261]: Discussed to maintain common flag for all registries through `*=...`

**Review by ilia1243 — APPROVED** [review]

File: kubemarine/core/defaults.py (hunk 2 above)
Comment (Imadzuma) [reply kubemarine/core/defaults.py]: bool(cluster) was fixed.
Enrichment changes only for registries, if they are enriched in new format. In other cases nothing changes.

File: kubemarine/patches/p3_reconfigure_registries.py (hunk 3 above)
Comment (Imadzuma) [reply kubemarine/patches/p3_reconfigure_registries.py:38]: 'containerdRegistriesConfig' for `prev_cluster` checks, that patch is not applied after manual changes in cluster.yaml. In that case user should manually run containerd reconfiguration, but patch should do nothing.
In other cases `cluster.inventory['services']['cri'] == prev_cluster.inventory['services']['cri']` handles the same cases, for this reason it doesn't matter what is used in that condition.

**Review by ilia1243 — APPROVED** [review]

**Pull request merged**